tensorflow/core/platform/tstring.h

```diff
@@ -225,7 +225,23 @@ class tstring {
   friend bool operator==(const std::string& a, const tstring& b);
   friend tstring operator+(const tstring& a, const tstring& b);
   friend std::ostream& operator<<(std::ostream& o, const tstring& str);
-  friend std::hash<tstring>;
+  #if !defined(__APPLE__)
```

Comment:
This looks pretty random, without understanding the underlying issue. Could you please investigate what the actual problem is?

Reply:
I tried to define a simplest class named `tstring` with `friend std::hash<tstring>` and it can be compiled! So it's confusing why the declaration here raises the error. I'm not sure if I can find the cause.

Reply:
@gharibian to take a look at this. 

From what I can tell, this is your error:
error: no template named 'hash'; did you mean 'std::hash'?
so I do not think there is an issue here.

Reply:
The removal may break things, as this hash is used for ordering of tstrings in stl containers.

The error messages you posted below seem to report the compilation error of the *modification( from std::hash -> struct hash. What was the original error message with this file *unaltered*?

Reply:
@gharibian The original error message is simply without the keyword `struct` (other information is the same).  When I first encountered the error, I searched suggestions on Stackoverflow and tried adding the keyword `struct` but without luck. I'll compile the code on a different mac to see if the problem persists.

Reply:
For us, this seems to be working on macos.
And given that its removal will break things, we cannot accept the change in this file.

Reply:
I tried building TF without CUDA using Xcode 11.4 and it's working.  It suggest the error may be a compiler bug of Xcode 9.4.1. Next I plan to update my compiler environment to Xcode 10.2 with CUDA 10.1. If that works, I'll revert this change.

Reply:
I think this is a compiler bug since I didn't meet the error using CUDA10.1 with XCode 10.1. However, a new bug https://github.com/tensorflow/tensorflow/issues/40260 prevented me from successful building TF with CUDA10.1. It means I have to use CUDA 10.0 with XCode 9.4.1 (CUDA10.1 does not support newer XCode) and this compiler bug needs the workaround here until https://github.com/tensorflow/tensorflow/issues/40260 is solved.

I understand both Apple and Nvidia started to abandon macOS as a CUDA compatible platform, and there are not many users insisting on using macOS to do deep learning. I still hope to offer a chance for the 'not many users' to use TF on macOS with CUDA. So I think we should make it work first before making it perfect.

Reply:
#40260 has been solved. Now I can revert this change.